Src/Base/AMReX_GpuFuse.H

```diff
@@ -0,0 +1,267 @@
+#include <AMReX_Arena.H>
+#include <AMReX_Vector.H>
+#include <AMReX_Box.H>
+#include <AMReX_GpuDevice.H>
+#include <AMReX_GpuLaunchGlobal.H>
+#include <cstring>
+#include <memory>
+
+namespace amrex {
+namespace Gpu {
+
+#ifdef AMREX_USE_CUDA
+
+typedef void (*Lambda1DLauncher)(char*,int);
+typedef void (*Lambda3DLauncher)(char*,int,int,int);
+typedef void (*Lambda4DLauncher)(char*,int,int,int,int);
+typedef void (*LambdaDtor)(char*);
+
+struct FuseHelper {
+    union LambdaLauncher {
+        Lambda1DLauncher L1D;
+        Lambda3DLauncher L3D;
+        Lambda4DLauncher L4D;
+    };
+    // Device function pointer to wrapper around Lambda's operator()
+    LambdaLauncher m_fp;
+    std::size_t m_offset; // This tells us where the lambda object is.
+    Box m_bx;
+    int m_N;
+};
+
+template <typename Lambda>
+AMREX_GPU_DEVICE
+void
+LaunchLambda1D (char* buf, int i)
+{
+    (*reinterpret_cast<Lambda*>(buf))(i);
+}
+
+template <typename Lambda>
+AMREX_GPU_DEVICE
+void
+LaunchLambda3D (char* buf, int i, int j, int k)
+{
+    (*reinterpret_cast<Lambda*>(buf))(i,j,k);
+}
+
+template <typename Lambda>
+AMREX_GPU_DEVICE
+void
+LaunchLambda4D (char* buf, int i, int j, int k, int n)
+{
+    (*reinterpret_cast<Lambda*>(buf))(i,j,k,n);
+}
+
+template <typename Lambda>
+void
+DestroyLambda (char* buf)
+{
+    reinterpret_cast<Lambda*>(buf)->~Lambda();
+}
+
+template <typename Lambda>
+std::size_t
+PutLambda (Lambda const& f, char* buf)
+{
+    new (buf) Lambda(f);
+    return Arena::align(sizeof(Lambda));
+}
+
+template <typename Lambda>
+void
+PutLambda1DLauncher (FuseHelper* helper)
+{
+    static Lambda1DLauncher fp = nullptr;
```

Comment:
This will not be thread-safe if we use OpenMP, right? We might want to have some sort of assert that prevents us from making that mistake.

Reply:
Yes, we need more assertions.  There is only one in the constructor.  But that's not enough especially because we are going to use a singleton.  I will add more assertions to `Register` and `Launch`.